### PR TITLE
Implement DIP1035 - system variables

### PR DESCRIPTION
**Pull request by dkorpel**
Make access of variables with `STC.system` a deprecation/error depending on `-preview=systemVariables`.
Does not implement inference of `@system` based on initializers yet, or checking against `void` initializing / array casting.

### CODE REVIEW

File: compiler/src/dmd/declaration.d [hunk 1]
```diff
@@ -542,6 +542,11 @@ extern (C++) abstract class Declaration : Dsymbol
         return (storage_class & STC.future) != 0;
     }
 
+    final extern(D) bool isSystem() const pure nothrow @nogc @safe
```
Comment (thewilsonator) [inline compiler/src/dmd/declaration.d:545]: Please update the header

Comment (dkorpel) [reply compiler/src/dmd/declaration.d:545]: It's `extern(D)`, do you think it should be exposed to the `extern(C++)` interface?

File: compiler/src/dmd/globals.h [hunk 2]
```diff
@@ -166,6 +166,8 @@ struct Param
     bool fix16997;               // fix integral promotions for unary + - ~ operators
                                  // https://issues.dlang.org/show_bug.cgi?id=16997
     FeatureState dtorFields;     // destruct fields of partially constructed objects
+    FeatureState systemVariables; // limit access to variables marked @system from @safe code
+
                                      // https://issues.dlang.org/show_bug.cgi?id=14246
```
[the remark this answers is not in the document]
Comment (dkorpel) [reply compiler/src/dmd/globals.h:171]: Good call

**Comment by dkorpel** [comment]
It's currently red because of deprecations when compiling druntime/phobos caused by https://github.com/dlang/druntime/pull/3007 which paints not only functions now, but also constants and variables `@system`.

File: compiler/src/dmd/declaration.d (hunk 1 above)
Comment (thewilsonator) [reply compiler/src/dmd/declaration.d:545]: Oh, I missed that. Thats a question for @ibuclaw or @kinke

Comment (RazvanN7) [reply compiler/src/dmd/declaration.d:545]: Nope, `extern(D)` does not need to be exposed to other backends.

**Review by RazvanN7 — COMMENTED** [review]
A list of suggestions and questions:

- declaring @system variable in a while/if condition and then using it should be tested.
- declaring @system parameters for functions should be tested
- is it allowed to take the address of a @system variable in @safe code? Anyway, a test with this should be added.
- is it allowed to pass a @system variable to a @safe function? 
- should we add a trait similar to getFunctionAttributes which gets the attributes of a variable since we now have attributes for variables?
- if you can declare a system parameter, is it possible to declare a template function that receives a system parameter?

File: druntime/src/core/sys/linux/ifaddrs.d [hunk 3]
```diff
@@ -22,7 +22,6 @@ version (linux):
 extern (C):
 nothrow:
 @nogc:
-@system:
```
[the remark this answers is not in the document]
Comment (dkorpel) [reply druntime/src/core/sys/linux/ifaddrs.d:25]: Remember [DIP1028](https://dlang.org/dips/1028) and [what happened to it](https://forum.dlang.org/post/pnjepgllfmgofwuegcqf@forum.dlang.org)?

Change my mind

I don't think I have to account for that anymore.

**Comment by dkorpel** [comment]
>  * declaring @System variable in a while/if condition and then using it should be tested.

Not allowed by the grammar
 
>  * declaring @System parameters for functions should be tested

Not allowed by the grammar

>  * is it allowed to take the address of a @System variable in @safe code?

No, then you access it

> Anyway, a test with this should be added.

Okay

>  * is it allowed to pass a @System variable to a @safe function?

Only if you're calling it from a `@system` or `@trusted` function

>  * should we add a trait similar to getFunctionAttributes which gets the attributes of a variable since we now have attributes for variables?

Probably, but not in this PR

>  * if you can declare a system parameter, is it possible to declare a template function that receives a system parameter?

You can't, and you can't

**Comment by dkorpel** [comment]
@RazvanN7 Anything else?

**Review by RazvanN7 — APPROVED** [review]
Looks good to me, however, I would prefer if @WalterBright merges this.

**Pull request merged**